docker-app/qfieldcloud/core/utils2/storage.py

```diff
@@ -378,7 +456,11 @@ def get_stored_package_ids(project_id: str) -> Set[str]:
 
 
 def delete_stored_package(project_id: str, package_id: str) -> None:
-    bucket = qfieldcloud.core.utils.get_s3_bucket()
     prefix = f"projects/{project_id}/packages/{package_id}/"
 
-    bucket.object_versions.filter(Prefix=prefix).delete()
+    if not re.match(r"^projects/[\w-]+/packages/\w+/$", prefix):
+        raise RuntimeError(
+            f"Suspicious S3 deletion on stored project package {project_id=} {package_id=}"
+        )
+
+    delete_by_prefix_permanently(prefix)
```

Comment:
(code style related, feel free to ignore as probably not worth debating this during the night :-) )

I would keep these checks about the key close to the actual S3 `delete` call. Since you moved the call to `delete_by_prefix_permanently()`, I'd also move the check there. Since it depends on the call, you could add a required `prefix_check_regex` argument, so that the pre-delete validation happens in exactly one place.

(personally I found it more clear to not factorize the call out, it was just a single function, but up to you)

Reply:
I like it more like with these two `delete_by_prefix_*` functions, because today I the `.delete()` was scattered around and unclear whether it causes permanent or versioned deletion, especially for people who are not familiar with `boto3`.

On the regex passed to the deleting function, the problem is  that the error message and the regex should be passed both, but I am not sure we want to do that.

My idea was to use these two functions internally in this module only, would it work for you if I prefix them with `_`? I will do that anyways.

Reply:
The only concern with splitting the check, is that it's unclear if the function expects the prefix to be sanitized already or not (looking at the code a bit too quickly and you think it's doing some sanitizing, yet you could still mess things up with a poorly formatted key for instance).

But really in the short term it makes no difference, just go for what you like most then we can improve afterwards

Reply:
Closing for now, will prefer not to change it now.